api/v1alpha1/etcdbackup_types.go

```diff
@@ -1,74 +1,33 @@
 package v1alpha1
 
 import (
-	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
-// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
-
 type EtcdBackupPhase string
 
 var (
 	EtcdBackupPhaseBackingUp EtcdBackupPhase = "BackingUp"
-	EtcdBackupPhaseUploading EtcdBackupPhase = "Uploading"
 	EtcdBackupPhaseCompleted EtcdBackupPhase = "Completed"
 	EtcdBackupPhaseFailed    EtcdBackupPhase = "Failed"
 )
 
-// EtcdClusterEndpoint holds an addressable endpoint for an etcd member.
-type EtcdClusterEndpoint struct {
-	// Port that is exposing the etcd client API for this member.
-	Port int `json:"port"`
-	// An IP address or DNS name of an endpoint.
-	Host string `json:"host"`
-	// Scheme to use for connecting to the host.
-	Scheme corev1.URIScheme `json:"scheme"`
-}
-
 // EtcdBackupDestination holds a storage location where an etcd backup will be placed.
-// At most one destination must be set.
 type EtcdBackupDestination struct {
-	// Local, when set, will copy the backup into a local volume on the pod taking the backup.
-	// +optional
-	Local *EtcdBackupDestinationLocal `json:"local,omitempty"`
-	// GCSBucket, when set, will push the backup to a Google Cloud Storage bucket.
-	// +optional
-	GCSBucket *EtcdBackupDestinationGCSBucket `json:"gcsBucket,omitempty"`
-}
-
-// EtcdBackupDestinationLocal describes a local directory into which to put the backup file. This is in the
-// filesystem of the pod running the operator. To persist this between pod restarts, ensure that path is
-// inside a mounted volume.
-type EtcdBackupDestinationLocal struct {
-	// Directory is an absolute filepath to a directory where backups will be placed.
-	Directory string `json:"path"`
-}
-
-// EtcdBackupDestinationGCSBucket describes a remote storage bucket on Google Cloud Storage, and the mechanisms used
-// to access this bucket.
-type EtcdBackupDestinationGCSBucket struct {
-	// BucketName is the name of the storage bucket.
-	//+kubebuilder:validation:MinLength=3
-	//+kubebuilder:validation:MaxLength=222
-	BucketName string `json:"bucketName"`
-	// Credentials holds the method of obtaining credentials that will be provided to the
-	// Google Cloud APIs in order to write backup data.
-	// +optional
-	Credentials *GoogleCloudCredentials `json:"credentials,omitempty"`
+	// ObjectURL is a URL of a file of a backup in object storage.
+	// The scheme of this URL should be gs:// or s3://.
+	ObjectURL string `json:"objectURL"`
```

Comment:
I've deviated from the design document, because I didn't understand why we shouldn't just specify the destination URL up front, which makes testing much simpler and also makes the backup API symmetrical with the restore API.

I will need to alter the EtcdBackupSchedule controller if I leave it like this, because it will need to uniquify the the backup objectURL for each backup resource that it creates.

Reply:
Done.

Reply:
Agreed to change the field name to `ObjectURLTemplate`, for now.